src/main/scala/is/hail/expr/Type.scala

```diff
@@ -228,6 +237,55 @@ sealed abstract class Type extends BaseType with Serializable {
   /*  Fundamental types are types that can be handled natively by RegionValueBuilder: primitive
       types, Array and Struct. */
   def fundamentalType: Type = this
+
+  def required: Boolean
+  def _typeCheck(a: Any): Boolean
+  final def typeCheck(a: Any): Boolean = (!required && a == null) || _typeCheck(a)
+
+  final def unary_!(): Type = {
+    this match {
+      case TBinary(req) => TBinary(!req)
+      case TBoolean(req) => TBoolean(!req)
+      case TInt32(req) => TInt32(!req)
+      case TInt64(req) => TInt64(!req)
+      case TFloat32(req) => TFloat32(!req)
+      case TFloat64(req) => TFloat64(!req)
+      case TString(req) => TString(!req)
+      case TGenotype(req) => TGenotype(!req)
+      case TCall(req) => TCall(!req)
+      case TAltAllele(req) => TAltAllele(!req)
+      case t: TArray => t.copy(required = !t.required)
+      case t: TSet => t.copy(required = !t.required)
+      case t: TDict => t.copy(required = !t.required)
+      case t: TVariant => t.copy(required = !t.required)
+      case t: TLocus => t.copy(required = !t.required)
+      case t: TInterval => t.copy(required = !t.required)
+      case t: TStruct => t.copy(required = !t.required)
+    }
+  }
+
+  final def isOfType(t: Type): Boolean = {
+    this match {
+      case TBinary(_) => t == TBinaryOptional || t == TBinaryRequired
+      case TBoolean(_) => t == TBooleanOptional || t == TBooleanRequired
+      case TInt32(_) => t == TInt32Optional || t == TInt32Required
+      case TInt64(_) => t == TInt64Optional || t == TInt64Required
+      case TFloat32(_) => t == TFloat32Optional || t == TFloat32Required
+      case TFloat64(_) => t == TFloat64Optional || t == TFloat64Required
+      case TString(_) => t == TStringOptional || t == TStringRequired
+      case TGenotype(_) => t == TGenotypeOptional || t == TGenotypeRequired
+      case TCall(_) => t == TCallOptional || t == TCallRequired
+      case TAltAllele(_) => t == TAltAlleleOptional || t == TAltAlleleRequired
+      case t2: TLocus => t == t2 || t == !t2
+      case t2: TVariant => t == t2 || t == !t2
+      case t2: TInterval => t == t2 || t == !t2
+      case t2: TStruct => t.isInstanceOf[TStruct] &&
+        t.asInstanceOf[TStruct].fields.zip(t2.fields).forall { case (f1: Field, f2: Field) => f1.typ.isOfType(f2.typ) }
+      case t2: TArray => t.isInstanceOf[TArray] && t.asInstanceOf[TArray].elementType.isOfType(t2.elementType)
+      case t2: TSet => t.isInstanceOf[TSet] && t.asInstanceOf[TSet].elementType.isOfType(t2.elementType)
+      case t2: TDict => t.isInstanceOf[TDict] && t.asInstanceOf[TDict].elementType.isOfType(t2.elementType)
```

Comment:
key type?

Reply:
I think using elementType provides the correct result, but splitting it out into a check on keyType and a check on valueType makes it more explicit and also maybe avoids a level of nesting, so I changed it.